internal/controllers/machinedeployment/machinedeployment_controller.go

```diff
@@ -282,6 +295,33 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 	return errors.Errorf("unexpected deployment strategy type: %s", md.Spec.Strategy.Type)
 }
 
+func (r *Reconciler) reconcileDelete(ctx context.Context, md *clusterv1.MachineDeployment) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	msList, err := r.getMachineSetsForDeployment(ctx, md)
```

Comment:
We might use this chance to explore using a controller runtime built-in deleteAllOf https://github.com/kubernetes-sigs/controller-runtime/blob/162a113134deee49b2c93abd9e35211dfe7783e6/pkg/client/interfaces.go#L79-L80, thougths?

Reply:
Same for ms/machines

Reply:
that might require bigger refactor so it's fine if we want to re-consider in a follow up

Reply:
I'm not sure if there is an option that we can use that covers exactly the MachineSets that we want to delete

I think currently we delete all MachineSets that:
* match the MD selector (except if the selector is empty, but we can just skip deletion entirely in that case)
* MachineSets without deletionTimestamp (although all MachineSets probably have the deletionTImestamp after the first reconcileDelete)

But then we also don't really have that many MachineSets for MD.

Maybe it's better to precisely delete exactly the MachineSets we want (with corresponding log lines) vs. using a DeleteAllOf and then only the server knows which MachineSets are deleted?